src/main/java/net/logstash/logback/util/ReusableJsonFormatterPool.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright 2013-2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package net.logstash.logback.util;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.lang.ref.SoftReference;
+import java.util.Deque;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentLinkedDeque;
+
+import net.logstash.logback.composite.CompositeJsonFormatter;
+
+import ch.qos.logback.core.spi.DeferredProcessingAware;
+
+/**
+ * @author brenuart
+ *
+ */
+public class ReusableJsonFormatterPool<Event extends DeferredProcessingAware> {
+
+    private final int minBufferSize;
+    
+    private final CompositeJsonFormatter<Event> formatterFactory;
+    
+    private SoftReference<Deque<ReusableJsonFormatter>> formatters = new SoftReference<>(null);
+    
+
+    public ReusableJsonFormatterPool(CompositeJsonFormatter<Event> formatterFactory, int minBufferSize) {
+        this.formatterFactory = Objects.requireNonNull(formatterFactory);
+        this.minBufferSize = minBufferSize;
+    }
+
+    public class ReusableJsonFormatter implements Closeable {
+        private final ReusableByteBuffer buffer;
+        private final CompositeJsonFormatter<Event>.JsonFormatter formatter;
+        
+        ReusableJsonFormatter(ReusableByteBuffer buffer, CompositeJsonFormatter<Event>.JsonFormatter formatter) {
+            this.buffer = buffer;
+            this.formatter = formatter;
+        }
+        
+        public ReusableByteBuffer getBuffer() {
+            return buffer;
+        }
+        
+        public void write(Event event) throws IOException {
+            this.formatter.writeEvent(event);
+        }
+        
+        @Override
+        protected void finalize() throws Throwable {
+            dispose();
+        }
```

Comment:
Hmm.  Finalizers are not that great, since they are expensive for the gc to execute.  I tend to avoid them.  Is there an explicit need for it here?

Reply:
The underlying JsonGenerator needs to be explicitly disposed for Jackson to release the associated resources (return buffers to its pool). But honestly I don't know what are the consequences if it is not explicitly disposed and garbage collected... Jackson pools its memory buffers a bit the same way as we do here: a buffer is removed from the pool when used and added back when released. The pool does not maintain reference to the buffers that are currently in use - so chances are that the only consequence of not explicitly disposing the JsonGenerator would be not to return the buffer to the pool and have a _cache miss_ next time...

The "problem" here is we don't know when a JsonFormatter is not needed anymore and can be disposed - we rely on the SoftReference that holds the pool (Deque). JsonFormatter will be disposed when the GC decides it needs more memory - reason why I relied on the finalize method. IMHO the overhead should not be too high - and should occur only when running low in memory.

Reply:
See https://github.com/logstash/logstash-logback-encoder/pull/631#issuecomment-914185865
Short version: implementation does not rely on `#finalize()` anymore.